src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -8,582 +8,198 @@
-    vcall = printer._vmerge(printer._hmerge([node.fencil.id + "("]), printer._indent(args), [");"])
-    call = printer._optimum(hcall, vcall)
-
-    body = printer._vmerge(*tmps, fencil, call)
-    return printer._vmerge(params, printer._indent(body), ["}"])
 
-
-PrettyPrinter.visit_FencilWithTemporaries = pformat_FencilWithTemporaries  # type: ignore
-
-
-# Main implementation
-def canonicalize_applied_lift(closure_params: list[str], node: ir.FunCall) -> ir.FunCall:
-    """
-    Canonicalize applied lift expressions.
-
-    Transform lift such that the arguments to the applied lift are only symbols.
-
-    >>> bool_type = ts.ScalarType(kind=ts.ScalarKind.BOOL)
-    >>> it_type = it_ts.IteratorType(position_dims=[], defined_dims=[], element_type=bool_type)
-    >>> expr = im.lift(im.lambda_("a")(im.deref("a")))(im.lift("deref")(im.ref("inp", it_type)))
-    >>> print(expr)
-    (↑(λ(a) → ·a))((↑deref)(inp))
-    >>> print(canonicalize_applied_lift(["inp"], expr))
-    (↑(λ(inp) → (λ(a) → ·a)((↑deref)(inp))))(inp)
-    """
-    assert cpm.is_applied_lift(node)
-    stencil = node.fun.args[0]  # type: ignore[attr-defined]  # ensured by is_applied lift
-    it_args = node.args
-    if any(not isinstance(it_arg, ir.SymRef) for it_arg in it_args):
-        closure_param_refs = collect_symbol_refs(node, as_ref=True)
-        assert not ({str(ref.id) for ref in closure_param_refs} - set(closure_params))
-        new_node = im.lift(
-            im.lambda_(*[im.sym(param.id) for param in closure_param_refs])(
-                im.call(stencil)(*it_args)
+    if extracted_fields:
+        tmp_stmts: list[itir.Stmt] = []
+
+        # for each extracted expression generate:
+        #  - one or more `Temporary` declarations (depending on whether the expression is a field
+        #    or a tuple thereof)
+        #  - one `SetAt` statement that materializes the expression into the temporary
+        for tmp_sym, tmp_expr in extracted_fields.items():
+            domain = tmp_expr.annex.domain
+
+            # TODO(tehrengruber): Implement. This happens when the expression for a combination
+            #  of an `if_` call with a tuple, e.g., `if_(cond, {a, b}, {c, d})`. As long as we are
+            #  able to eliminate all tuples, e.g., by propagating the scalar ifs to the top-level
+            #  of a SetAt, the CollapseTuple pass will eliminate most of this cases.
+            if isinstance(domain, tuple):
+                flattened_domains: tuple[domain_utils.SymbolicDomain] = (
+                    next_utils.flatten_nested_tuple(domain)  # type: ignore[assignment]  # mypy not smart enough
+                )
+                if not all(d == flattened_domains[0] for d in flattened_domains):
+                    raise NotImplementedError(
+                        "Tuple expressions with different domains is not " "supported yet."
+                    )
+                domain = flattened_domains[0]
```

Comment:
Optional suggestion for an alternative (cleaner?) implementation of this (assuming  `SymbolicDomain` is a hashable frozen dataclass):
```suggestion
                flattened_domains: tuple[domain_utils.SymbolicDomain] = set(
                    next_utils.flatten_nested_tuple(domain)
                )
                domain = flattened_domains.pop()
                if flattened_domains:
                    raise NotImplementedError(
                        "Tuple expressions with different domains is not " "supported yet."
                    )
```

Reply:
SymbolicDomain contains a dict and is not hashable.